File: docs/superpowers/specs/bar-preview.py
#!/usr/bin/env python3
"""clave sidebar — the LOCKED visual design, rendered.

    python3 docs/superpowers/specs/bar-preview.py

An ILLUSTRATION of the sidebar design ratified 2026-07-25 — not its authority.
`2026-07-25-sidebar-visual-design-lock.md` in this directory is authoritative
for every ruling, number and rationale; where this script and that document
disagree, the document wins and this script is the bug.

It is deliberately a standalone renderer with no clave dependency, so it runs in
any checkout and any terminal without building the wasm plugin, and it asserts
its own core invariant: every row is exactly COLS display cells, measured in
cells rather than code points.

FOLLOW-UP (tracked): this preview duplicates geometry that `compose_row` will
own once S5/S6/S8 land. At that point it should become a Rust example driven by
the real constants, so a code change that moves a column breaks the preview
instead of silently diverging from it. Until then, treat the prose spec as
authoritative and this file as its illustration.

GLYPH RULE — load-bearing. Every glyph below is spelled as a \\U000xxxxx escape,
never as a literal character. During the design rounds, literal glyphs were
silently lost in transit twice; the first time we misdiagnosed it as missing
font coverage and nearly constrained the whole design to one Unicode plane on
the strength of it. Escapes survive every tool in the chain. The same rule
applies to the Rust source: write '\\u{f062c}', not the glyph.
"""

import re
import unicodedata

# \u2500\u2500 ANSI \u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500
R, BOLD = "\x1b[0m", "\x1b[1m"


def rgb(h):
    h = h.lstrip("#")
    return int(h[0:2], 16), int(h[2:4], 16), int(h[4:6], 16)


def fg(h):
    r, g, b = rgb(h)
    return f"\x1b[38;2;{r};{g};{b}m"


def bg(h):
    r, g, b = rgb(h)
    return f"\x1b[48;2;{r};{g};{b}m"


def mix(h, toward, t):
    """Blend `h` toward `toward` by t. Unselected rows are rendered at
    FADE toward the bar background — selection by recession (§4)."""
    a, b_, c = rgb(h)
    d, e, f_ = rgb(toward)
    return "#%02X%02X%02X" % (round(a + (d - a) * t), round(b_ + (e - b_) * t),
                              round(c + (f_ - c) * t))


# \u2500\u2500 locked constants \u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500
COLS = 44                  # expanded bar width
FADE = 0.25                # unselected rows, blended toward BASE
BASE = "#1F1F28"           # kanagawa sumiInk1 — the bar background
RULE_INK = "#DCD7BA"       # kanagawa fujiWhite — the status rule
SEL_BG = "#2D4F67"         # kanagawa waveBlue2 — the selected row
CHIP_INK = "#16161D"       # kanagawa sumiInk0 — text ON a title chip
LCAP, RCAP = "\ue0b6", "\ue0b4"   # powerline half-circle thick (U+E0B6/E0B4)
ELLIPSIS = "\u2026"
RULE = "\u2502"                    # box drawings light vertical

TITLE_W, REPO_W = 7, 7

# The repo palette: 8 kanagawa hues, allocated round-robin and keyed by repo
# root, so one repo is one colour everywhere, forever. Title chips draw from
# the SAME palette but are keyed per-title WITHIN a repo, so two tabs of the
# same repo never share a chip. See the prose spec §5.
PALETTE = [("#7E9CD8", "crystalBlue"), ("#98BB6C", "springGreen"),
           ("#E6C384", "carpYellow"), ("#E46876", "waveRed"),
           ("#957FB8", "oniViolet"), ("#7AA89F", "waveAqua2"),
           ("#FFA066", "surimiOrange"), ("#D27E99", "sakuraPink")]

# Cell 1 of the gutter. The COLOUR is the status; the shape never varies
# except for the two terminal states.
STATUS = {"working": ("\u25cf", "#FF9E3B"), "done": ("\u25cf", "#98BB6C"),
          "unread": ("\u25cf", "#E46876"), "idle": ("\u25cf", "#54546D"),
          "dormant": ("\u25cc", "#54546D"), "stale": ("\u2717", "#E82424")}

# Cell 3 of the gutter — the context battery (S7). Colour is the magnitude
# ramp, green through red. A plain terminal tab shows the console mark here
# instead, because a terminal has no context window.
BATTERY = [("\U000f0079", "#98BB6C"), ("\U000f007e", "#98BB6C"),
           ("\U000f007c", "#E6C384"), ("\U000f007b", "#FFA066"),
           ("\U000f007a", "#E46876")]
CONSOLE = "\U000f018d"                 # nf-md-console

# Cell 4 — provenance. Tinted with the REPO ink, so it matches the repo name.
# A main checkout deliberately shows NOTHING: no terminal tool in the survey
# marks the default branch with a glyph, and blanking it makes the two marked
# states actually mean something.
PROVENANCE = {"worktree": "\U000168c2",   # bamum tree
              "branch": "\U000f062c",     # nf-md-source_branch (lazygit's)
              "main": " "}


def cells(s):
    """Terminal CELLS occupied by `s`, not code points.

    `len()` counts scalars: a wide (East-Asian W/F) glyph occupies two columns
    and a combining mark none, so code-point arithmetic silently misaligns
    every row to its right. This is the same hazard the dossier records against
    the renderer's `str::chars()` clamp. Rust must use `unicode-width` — the
    exact version zellij lays its grid with — not `chars().count()`.
    """
    n = 0
    for ch in s:
        if unicodedata.combining(ch):
            continue
        n += 2 if unicodedata.east_asian_width(ch) in ("W", "F") else 1
    return n


def clamp(s, w):
    """A fixed-width column, measured in CELLS: truncate when long, pad when
    short. Padding is what makes fields line up vertically down the bar."""
    if w <= 0:
        return ""
    if cells(s) <= w:
        return s + " " * (w - cells(s))
    out = ""
    for ch in s:
        if cells(out) + cells(ch) > w - 1:
            break
        out += ch
    return out + ELLIPSIS + " " * (w - cells(out) - 1)


def render(row, selected=False):
    """One row, exactly COLS columns wide.

    The gutter is POSITION-LOCKED: every cell is one column and renders a
    space when its glyph is absent, so a missing glyph can never shift the
    text. The cap columns are reserved on EVERY row for the same reason —
    the selected row must not sit one column right of its neighbours.
    """
    fade = 0.0 if selected else FADE
    o = bg(SEL_BG) if selected else ""

    def ink(h):
        return fg(mix(h, BASE, fade))

    is_term = row.get("terminal")
    repo_ink = row.get("repo_ink", "#54546D")

    out = (fg(SEL_BG) + LCAP + o) if selected else " "          # col 1
    if is_term:
        out += o + " "                                          # col 2
        out += o + " " + ink(RULE_INK) + RULE + o + " "         # cols 3-5
        out += o + ink("#54546D") + CONSOLE + o                 # col 6
        out += o + "  "                                         # cols 7-8
    else:
        sg, si = STATUS[row["status"]]
        bg_, bi = BATTERY[row["battery"]]
        out += o + ink(si) + sg + o                             # col 2
        out += o + " " + ink(RULE_INK) + RULE + o + " "         # cols 3-5
        out += o + ink(bi) + bg_ + o                            # col 6
        out += o + " "                                          # col 7
        pv = PROVENANCE[row["provenance"]]
        out += o + (ink(repo_ink) + pv + o if pv.strip() else " ")   # col 8
    out += o + " "                                              # col 9

    body = COLS - 9 - 1 - 1          # minus gutter, right margin, right cap
    summary_w = body - TITLE_W - REPO_W - 2
    if is_term:
        out += o + ink("#717C7C") + clamp(row["name"], body) + o
    else:
        title = row.get("title")
        if title:
            out += bg(mix(row["title_ink"], BASE, fade)) + fg(CHIP_INK) \
                + clamp(title, TITLE_W) + R + o
        else:
            out += o + " " * TITLE_W
        out += o + " "
        out += ink(repo_ink) + clamp(row["repo"], REPO_W) + o
        out += o + " "
        out += (ink("#DCD7BA") if fade else "") \
            + clamp(row["summary"], summary_w) + o
    out += o + " " + R                                          # right margin
    out += (fg(SEL_BG) + RCAP + R) if selected else " "         # col 44
    return out


# \u2500\u2500 a sample fleet \u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500
INK = {"clave": PALETTE[0][0], "dotfiles": PALETTE[1][0],
       "api-svc": PALETTE[2][0], "infra": PALETTE[4][0], "webapp": PALETTE[5][0]}

FLEET = [
    dict(status="unread", battery=1, provenance="main", repo="dotfiles",
         repo_ink=INK["dotfiles"], title=None,
         summary="I just passed the spec over"),
    dict(status="working", battery=3, provenance="worktree", repo="clave",
         repo_ink=INK["clave"], title="S6-GUT", title_ink=PALETTE[5][0],
         summary="picking the gutter set", selected=True),
    dict(status="working", battery=2, provenance="branch", repo="api-svc",
         repo_ink=INK["api-svc"], title="API-GW", title_ink=PALETTE[3][0],
         summary="retry budget audit"),
    dict(status="done", battery=0, provenance="main", repo="clave",
         repo_ink=INK["clave"], title="UX-DOC", title_ink=PALETTE[6][0],
         summary="there was a stale anchor"),
    dict(terminal=True, name="Tab #16"),
    dict(status="stale", battery=4, provenance="worktree", repo="clave",
         repo_ink=INK["clave"], title="KDL-GRD", title_ink=PALETTE[7][0],
         summary="the real parser test"),
    dict(status="dormant", battery=0, provenance="worktree", repo="clave",
         repo_ink=INK["clave"], title=None, summary="worktree doctor work"),
    dict(status="dormant", battery=1, provenance="branch", repo="infra",
         repo_ink=INK["infra"], title="CFG", title_ink=PALETTE[1][0],
         summary="config plumbing"),
    dict(status="idle", battery=2, provenance="branch", repo="webapp",
         repo_ink=INK["webapp"], title="API-V2", title_ink=PALETTE[2][0],
         summary="cutover plan review"),
]

DIMTXT = fg("#717C7C")


def bar(rows, label):
    ruler = "".join(str((i + 1) % 10) for i in range(COLS))
    print(f"\n  {BOLD}{label}{R}")
    print(f"  {DIMTXT}{ruler}\n  \u250c{"\u2500" * COLS}\u2510{R}")
    for r in rows:
        line = render(r, r.get("selected", False))
        # The lock doc CLAIMS every row is exactly COLS cells. Prove it rather
        # than asserting it in prose: strip the SGR sequences and measure the
        # remainder in display cells. A miscounted glyph fails the preview
        # loudly instead of shipping a ragged bar.
        width = cells(re.sub(r"\x1b\[[0-9;]*m", "", line))
        assert width == COLS, f"row is {width} cells, expected {COLS}: {r!r}"
        print(f"  {DIMTXT}\u2502{R}{line}{DIMTXT}\u2502{R}")
    print(f"  {DIMTXT}\u2514{"\u2500" * COLS}\u2518{R}")


if __name__ == "__main__":
    print(f"\n{BOLD}{"\u2550" * 78}\nclave sidebar — locked visual design "
          f"(2026-07-25)\n{"\u2550" * 78}{R}")
    bar(FLEET, f"expanded — {COLS} columns")

    print(f"""
  {DIMTXT}COLUMN MAP
     1      left cap   — powerline half-circle, selected row only
     2      status     — colour IS the state
     3      space
     4      rule       — U+2502 in fujiWhite, separates status from battery
     5      space
     6      battery    — context level (S7); console mark on a terminal tab
     7      space
     8      provenance — tinted with the repo ink; BLANK for a main checkout
     9      space
    10-16   title      — filled chip, dark text; blank when never renamed
    17      space
    18-24   repo       — tinted text, one colour per repo forever
    25      space
    26-42   summary
    43      right margin
    44      right cap  — selected row only

  Cap columns are reserved on EVERY row so the selected row does not shift
  one column right of its neighbours. Verified: title starts at column 10
  whether or not the row is selected.{R}""")

    print(f"\n  {BOLD}palette — 8 kanagawa hues, round-robin{R}\n")
    for i, (h, n) in enumerate(PALETTE):
        print(f"   {DIMTXT}{i}{R} {fg(h)}\u2588\u2588\u2588\u2588{R}  "
              f"{fg(h)}repo-name{R}   {bg(h)}{fg(CHIP_INK)} TITLE {R}   "
              f"{DIMTXT}{h}  {n}{R}")
    print()

File: crates/clave-bar/examples/bar-preview.rs
//! clave sidebar — the LOCKED visual design, rendered.
//!
//!     cargo run -p clave-bar --example bar-preview
//!
//! An ILLUSTRATION of the sidebar design ratified 2026-07-25 — not its
//! authority. `docs/superpowers/specs/2026-07-25-sidebar-visual-design-lock.md`
//! is authoritative for every ruling, number and rationale; where this example
//! and that document disagree, the document wins and this file is the bug.
//!
//! Every row below comes from `clave_bar::render::render_rows` — the same
//! function the plugin renders with. That is the whole point of this file
//! existing in Rust: the Python original it replaces was a SECOND
//! implementation of the same geometry, and two renders of one design diverge.
//! Move a column in `render.rs` and this preview changes with it, or the width
//! assertion below fails; it can no longer drift silently.
//!
//! It still asserts its own core invariant, as the Python did: every row is
//! exactly `DESIGN_COLS` display CELLS, measured in cells rather than code
//! points.
//!
//! GLYPH RULE — load-bearing (lock §5.4). Every glyph below is a `\u{...}`
//! escape, never a literal character. During the design rounds literal glyphs
//! were silently lost in transit twice; the first loss was misdiagnosed as
//! missing font coverage and nearly constrained the whole design to one Unicode
//! plane. Escapes survive every tool in the chain.

use clave_bar::render::{
    CHIP_INK, COLLAPSED_DESIGN_COLS, DESIGN_COLS, PALETTE, Provenance, RESET, Rgb, Row, RowContent,
    RowStatus, TermStatus, Widths, display_cells, render_rows, strip_sgr,
};

#[path = "shared/showcase_fixture.rs"]
mod showcase_fixture;
use showcase_fixture::{API_SVC, CLAVE, DOTFILES, INFRA, WEBAPP, agent, showcase, terminal};

const BOLD: &str = "\u{1b}[1m";
/// The preview's own chrome — frame, ruler and column map. Not part of the
/// design; it is the graph paper the design is drawn on.
const DIM: Rgb = Rgb(0x71, 0x7C, 0x7C);

fn fleet() -> Vec<Row> {
    let mut rows = vec![
        agent(
            RowStatus::NeedsYou,
            Some((1, 15_000)),
            Provenance::Main,
            "dotfiles",
            DOTFILES,
            None,
            "I just passed the spec over",
        ),
        agent(
            RowStatus::Working,
            Some((3, 45_000)),
            Provenance::Worktree,
            "clave",
            CLAVE,
            Some(("S6-GUT", 5)),
            "picking the gutter set",
        ),
        agent(
            RowStatus::Working,
            Some((2, 30_000)),
            Provenance::Branch,
            "api-svc",
            API_SVC,
            Some(("API-GW", 3)),
            "retry budget audit",
        ),
        agent(
            RowStatus::Done,
            Some((0, 4_000)),
            Provenance::Main,
            "clave",
            CLAVE,
            Some(("UX-DOC", 6)),
            "there was a stale anchor",
        ),
        terminal(
            "Tab #16",
            TermStatus::Running,
            Provenance::Worktree,
            Some(("clave", CLAVE)),
            "cargo test --workspace",
        ),
        // The nothing-known-yet default: a freshly opened tab, at its prompt,
        // before any pane fact has arrived.
        Row {
            content: RowContent::terminal("Tab #3"),
            selected: false,
            dormant: false,
        },
        agent(
            RowStatus::Stale,
            Some((4, 66_000)),
            Provenance::Worktree,
            "clave",
            CLAVE,
            Some(("KDL-GRD", 7)),
            "the real parser test",
        ),
        agent(
            RowStatus::Dormant,
            Some((0, 9_000)),
            Provenance::Worktree,
            "clave",
            CLAVE,
            None,
            "worktree doctor work",
        ),
        agent(
            RowStatus::Dormant,
            Some((1, 22_000)),
            Provenance::Branch,
            "infra",
            INFRA,
            Some(("CFG", 1)),
            "config plumbing",
        ),
        agent(
            RowStatus::Idle,
            Some((2, 38_000)),
            Provenance::Branch,
            "webapp",
            WEBAPP,
            Some(("API-V2", 2)),
            "cutover plan review",
        ),
    ];
    rows[1].selected = true;
    rows
}

/// `cols`/`widths` are parameters (task 1.5, LEDGER D16) so the same framed
/// box — ruler, border, per-row width assertion — draws both the ratified
/// expanded render and every collapsed candidate. Behaviour at the original
/// call site (`cols = DESIGN_COLS`, `widths = Widths::EXPANDED`) is unchanged;
/// this is a signature widening, not a rewrite.
fn bar(rows: &[Row], cols: usize, widths: Widths, label: &str) {
    let ruler: String = (0..cols)
        .map(|i| char::from_digit(((i + 1) % 10) as u32, 10).unwrap())
        .collect();
    let dim = DIM.fg();
    let rule = "\u{2500}".repeat(cols);
    println!("\n  {BOLD}{label}{RESET}");
    println!("  {dim}{ruler}\n  \u{250c}{rule}\u{2510}{RESET}");
    for (line, row) in render_rows(rows, cols, rows.len(), widths).iter().zip(rows) {
        // The lock CLAIMS every row is exactly `cols` cells. Prove it rather
        // than asserting it in prose: strip the SGR sequences and measure the
        // remainder in display cells. A miscounted glyph fails the preview
        // loudly instead of shipping a ragged bar.
        let width = display_cells(&strip_sgr(line));
        assert_eq!(width, cols, "row is {width} cells: {row:?}");
        println!("  {dim}\u{2502}{RESET}{line}{dim}\u{2502}{RESET}");
    }
    println!("  {dim}\u{2514}{rule}\u{2518}{RESET}");
}

/// Bare rows, nothing else on stdout. The width assertion still runs — a
/// screenshot of a ragged bar would be worse than no screenshot.
fn print_showcase() {
    let rows = showcase();
    for (line, row) in render_rows(&rows, DESIGN_COLS, rows.len(), Widths::EXPANDED)
        .iter()
        .zip(&rows)
    {
        let width = display_cells(&strip_sgr(line));
        assert_eq!(width, DESIGN_COLS, "row is {width} cells: {row:?}");
        println!("{line}");
    }
}

fn main() {
    // `--showcase` prints the screenshot frame and exits: the chrome below is
    // for reading the design, not for showing it.
    if std::env::args().any(|a| a == "--showcase") {
        print_showcase();
        return;
    }
    let dim = DIM.fg();
    let bar_rule = "\u{2550}".repeat(78);
    println!(
        "\n{BOLD}{bar_rule}\nclave sidebar \u{2014} locked visual design \
         (2026-07-25)\n{bar_rule}{RESET}"
    );
    bar(
        &fleet(),
        DESIGN_COLS,
        Widths::EXPANDED,
        &format!("expanded \u{2014} {DESIGN_COLS} columns"),
    );

    column_map();

    println!("\n  {BOLD}palette \u{2014} 8 kanagawa hues, round-robin{RESET}\n");
    for (i, (hue, name)) in PALETTE.iter().enumerate() {
        let (fg, bg, hex) = (hue.fg(), hue.bg(), hue.hex());
        let chip = CHIP_INK.fg();
        println!(
            "   {dim}{i}{RESET} {fg}\u{2588}\u{2588}\u{2588}\u{2588}{RESET}  \
             {fg}repo-name{RESET}   {bg}{chip} TITLE {RESET}   {dim}{hex}  {name}{RESET}"
        );
    }
    println!();

    collapsed();
}

/// Where each field starts and ends, for a profile at a given width — the same
/// arithmetic `render_row` lays the row out with, so the map below cannot
/// describe a layout the renderer does not produce.
fn field_spans(w: Widths, cols: usize) -> [(usize, usize); 4] {
    // The battery cell opens at column 6 in both profiles and runs as wide as
    // the profile gives it — one column for the glyph, four for the count (#105).
    let battery = (6, 5 + w.battery.width());
    let title = (w.gutter() + 1, w.gutter() + w.title);
    let repo = (title.1 + 2, title.1 + 1 + w.repo);
    let summary = (repo.1 + 2, cols - 2);
    [battery, title, repo, summary]
}

fn span(s: (usize, usize)) -> String {
    if s.0 == s.1 {
        format!("{}", s.0)
    } else {
        format!("{}-{}", s.0, s.1)
    }
}

/// The column map, DERIVED from the `Widths` profiles rather than restated in
/// prose. Review finding 8 flagged the old hard-coded table as the one part of
/// this preview that could still silently diverge from the renderer — and it is
/// the part a reader trusts most. Moving a column now moves these numbers too.
fn column_map() {
    let dim = DIM.fg();
    let (x, c) = (Widths::EXPANDED, Widths::COLLAPSED);
    let (xs, cs) = (
        field_spans(x, DESIGN_COLS),
        field_spans(c, COLLAPSED_DESIGN_COLS),
    );

    println!(
        "\n  {dim}COLUMN MAP \u{2014} everything left of the battery is IDENTICAL in both
  states (D16); the battery cell is 4 cells expanded and 1 collapsed (#105), and
  title, repo and summary follow it. summary = cols \u{2212} gutter \u{2212} 4 \u{2212} title \u{2212} repo,
  and it is the ONLY flexible column (D9), so the wider battery cell was paid for
  out of it. The cells between the named ones are single spaces."
    );
    println!(
        "\n  {:<13}{:<11}{:<12}what it carries",
        "field", "expanded", "collapsed"
    );
    for (field, xa, ca, note) in [
        (
            "left cap",
            span((1, 1)),
            span((1, 1)),
            "powerline half-circle, selected row only",
        ),
        (
            "status",
            span((2, 2)),
            span((2, 2)),
            "the COLOUR is the state",
        ),
        (
            "rule",
            span((4, 4)),
            span((4, 4)),
            "U+2502 in fujiWhite, splits status from battery",
        ),
        (
            "battery",
            span(xs[0]),
            span(cs[0]),
            "token count expanded, ramp glyph collapsed (S7, #105)",
        ),
        (
            "provenance",
            span((x.gutter() - 1, x.gutter() - 1)),
            span((c.gutter() - 1, c.gutter() - 1)),
            "the repo ink; BLANK for a main checkout",
        ),
        (
            "title",
            span(xs[1]),
            span(cs[1]),
            "filled chip, dark text; blank when never renamed",
        ),
        (
            "repo",
            span(xs[2]),
            span(cs[2]),
            "tinted text, one colour per repo forever",
        ),
        (
            "summary",
            span(xs[3]),
            span(cs[3]),
            "flexes; no ellipsis at \u{2264}4 cells (D18)",
        ),
        (
            "right margin",
            span((DESIGN_COLS - 1, DESIGN_COLS - 1)),
            span((COLLAPSED_DESIGN_COLS - 1, COLLAPSED_DESIGN_COLS - 1)),
            "",
        ),
        (
            "right cap",
            span((DESIGN_COLS, DESIGN_COLS)),
            span((COLLAPSED_DESIGN_COLS, COLLAPSED_DESIGN_COLS)),
            "selected row only",
        ),
    ] {
        println!(
            "{}",
            format!("  {field:<13}{xa:<11}{ca:<12}{note}").trim_end()
        );
    }
    println!(
        "
  Cap columns are reserved on EVERY row so the selected row does not shift one
  column right of its neighbours. Verified: the chip starts at its profile's
  first text column whether or not the row is selected.{RESET}"
    );
}

/// The chosen collapsed profile (task 1.5 / LEDGER D17), rendered the same
/// way as the expanded section above it: `bar()` draws the ruler, the framed
/// box and the per-row width assertion for either profile identically — the
/// candidates comparison this replaces has served its purpose, Ollie picked
/// `(title 7, repo 3)` at 30 columns by looking at it.
fn collapsed() {
    let dim = DIM.fg();

    let widths = Widths::COLLAPSED;
    let cols = COLLAPSED_DESIGN_COLS;
    // Derived, not hard-coded: the same arithmetic `render_row` uses internally
    // once `cols` clears the floor (`gutter + 4 + title + repo` — LEDGER D12's
    // arithmetic, generalised to a profile by D16 and to a variable-width battery
    // cell by #105), so this number is guaranteed to match what the box below
    // actually shows.
    let summary_w = cols.saturating_sub(widths.min_intact_cols());
    bar(
        &fleet(),
        cols,
        widths,
        &format!(
            "collapsed \u{2014} {cols} columns \u{2014} title {}, repo {}, summary {}",
            widths.title, widths.repo, summary_w
        ),
    );

    let separation = DESIGN_COLS - cols;
    println!("\n  {dim}separation from {DESIGN_COLS}: {separation}{RESET}");
    println!(
        "  {dim}LEDGER D39: the separation is now a DESIGN number, not a \
         tolerance. Both widths are declared in the layout and zellij switches \
         between them, so nothing has to tell the two apart by measuring \
         \u{2014} what {separation} columns buys is a visibly different bar.{RESET}"
    );
}
